src/librustc_metadata/schema.rs

```diff
@@ -155,6 +155,14 @@ pub enum LazyState {
     Previous(NonZeroUsize),
 }
 
+// FIXME(#59875) `Lazy!(T)` replaces `Lazy<T>`, passing the `Meta` parameter
+// manually, instead of relying on the default, to get the correct variance.
+// Only needed when `T` itself contains a parameter (e.g. `'tcx`).
+macro_rules! Lazy {
+    ([$T:ty]) => {Lazy<[$T], usize>};
+    ($T:ty) => {Lazy<$T, ()>};
+}
```

Comment:
This macro is pretty ugly. So is just `Lazy<T>` invariant in `T` even with the `Meta` type parameter?

Where do you need the co-variance?

Reply:
The default has a projection in it, which variance computation doesn't normalize AFAIK.

I need the variance because e.g. `Lazy<Ty<'tcx>>` is more like `Lazy<for<'tcx> Ty<'tcx>>` - you can decode it as any lifetime (that you happen to have a decoder for, that is).
This allows holding `CrateRoot<'static>` instead of `CrateRoot<'tcx>` in `CrateMetadata`.

If we try hard enough, we might be able to make `CrateMetadata<'tcx>` work, but it might require using a trait object instead of `Provider`s for non-local crates. Or some unsafe code.